src/test/FreeC/Backend/Coq/Converter/FuncDecl/Rec/WithHelpersTests.hs

```diff
@@ -501,6 +501,83 @@ testConvertRecFuncDeclWithHelpers = context "with helper functions" $ do
           ++ " := t >>= (fun (t_0 : Tree Shape Pos a) =>"
           ++ "      @height_0 Shape Pos a t_0)."
 
+  it "translates recursive functions with one strict argument correctly"
+    $ shouldSucceedWith
+    $ do
+        "Integer"        <- defineTestTypeCon "Integer" 0
+        "succ"           <- defineTestFunc "succ" 1 "Integer -> Integer"
+        "List"           <- defineTestTypeCon "List" 1
+        ("nil" , "Nil" ) <- defineTestCon "Nil" 0 "forall a. List a"
+        ("cons", "Cons") <- defineTestCon "Cons"
+                                          2
+                                          "forall a. a -> List a -> List a"
+        "length" <- defineTestFunc "length" 1 "forall a. List a -> Integer"
+        shouldConvertWithHelpersTo
+            [ "length @a !(xs :: List a) :: Integer = case xs of {"
+              ++ "    Nil        -> 0;"
+              ++ "    Cons x xs' -> succ (length @a xs')"
+              ++ "  }"
+            ]
+          $  "(* Helper functions for length *) "
+          ++ "Fixpoint length_0 (Shape : Type) (Pos : Shape -> Type)"
+          ++ "  {a : Type} (xs : List Shape Pos a) {struct xs}"
+          ++ " := match xs with"
+          ++ "    | nil        => pure 0%Z"
+          ++ "    | cons x xs' => succ Shape Pos"
+          ++ "         (xs' >>= (fun (xs'_0 : List Shape Pos a) =>"
+          ++ "           @length_0 Shape Pos a xs'_0))"
+          ++ "    end. "
+          ++ "Definition length (Shape : Type) (Pos : Shape -> Type) {a : Type}"
+          ++ "  (xs : List Shape Pos a)"
+          ++ "  : Free Shape Pos (Integer Shape Pos)"
+          ++ " := @length_0 Shape Pos a xs."
+
+  it
+      "translates recursive functions with strict and non-strict arguments correctly"
+    $ shouldSucceedWith
+    $ do
+        "List"           <- defineTestTypeCon "List" 1
+        ("nil" , "Nil" ) <- defineTestCon "Nil" 0 "forall a. List a"
+        ("cons", "Cons") <- defineTestCon "Cons"
+                                          2
+                                          "forall a. a -> List a -> List a"
+        "interleave" <- defineTestFunc "interleave"
+                                       2
+                                       "forall a. List a -> List a -> List a"
+        shouldConvertWithHelpersTo
+            [ "interleave @a (xs :: List a) !(ys :: List a) :: List a ="
```

Comment:
I think it would be also interesting to have an example where the bang pattern precedes the decreasing argument. Such an example could reveal a problem in the generation of the `struct` annotation.